frame/election-provider-multi-phase/src/lib.rs

```diff
@@ -727,6 +766,72 @@ pub mod pallet {
 
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
+		/// Submit a solution for the signed phase.
+		///
+		/// The dispatch origin fo this call must be __signed__.
+		///
+		/// The solution is potentially queued, based on the claimed score and processed at the end
+		/// of the signed phase.
+		///
+		/// A deposit is reserved and recorded for the solution. Based on the outcome, the solution
+		/// might be rewarded, slashed, or get all or a part of the deposit back.
+		///
+		/// # <weight>
+		/// Queue size must be provided as witness data.
+		/// # </weight>
+		#[pallet::weight(T::WeightInfo::submit(*witness_data))]
+		pub fn submit(
+			origin: OriginFor<T>,
+			solution: RawSolution<CompactOf<T>>,
+			witness_data: u32,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+
+			// ensure witness data is correct.
+			ensure!(
+				witness_data >= <SignedSubmissions<T>>::decode_len().unwrap_or_default() as u32,
+				Error::<T>::SignedInvalidWitness,
+			);
+
+			// ensure solution is timely.
+			ensure!(Self::current_phase().is_signed(), Error::<T>::PreDispatchEarlySubmission);
+
+			// NOTE: this is the only case where having separate snapshot would have been better
+			// because could do just decode_len. But we can create abstractions to do this.
+
+			// build size. Note: this is not needed for weight calc, thus not input.
+			// defensive-only: if phase is signed, snapshot will exist.
+			let size = Self::snapshot_metadata().unwrap_or_default();
+
+			ensure!(
+				Self::feasibility_weight_of(&solution, size) < T::SignedMaxWeight::get(),
+				Error::<T>::SignedTooMuchWeight,
+			);
+
+			// ensure solution claims is better.
+			let mut signed_submissions = Self::signed_submissions();
+			let ejected_a_solution = signed_submissions.len()
+				== T::SignedMaxSubmissions::get().saturated_into::<usize>();
+			let index = Self::insert_submission(&who, &mut signed_submissions, solution, size)
+				.ok_or(Error::<T>::SignedQueueFull)?;
+
+			// collect deposit. Thereafter, the function cannot fail.
+			// Defensive -- index is valid.
```

Comment:
Why is the defensive choice in both of these cases to use `default`?

Just return an error, and avoid anyone mistakenly writing code which might otherwise work with a default value or whatever.

Reply:
The idea was just to avoid creating errors which aren't ever realistically going to be returned. That's fine, though; returning errors already meant that we demonstrated the need to improve a benchmark. 29d9242c5.